docker/elasticsearch-setup/create-indices.sh

```diff
@@ -5,71 +5,96 @@ set -e
+}
+
+# create indices for ES OSS (AWS)
 function create_datahub_usage_event_aws_elasticsearch() {
-  if [[ -z "$INDEX_PREFIX" ]]; then
-    PREFIX=''
-  else
-    PREFIX="${INDEX_PREFIX}_"
-  fi
+  create_if_not_exists "_opendistro/_ism/policies/${PREFIX}datahub_usage_event_policy" aws_es_ism_policy.json
+  create_if_not_exists "_template/${PREFIX}datahub_usage_event_index_template" aws_es_index_template.json
 
-  if [ $(curl -o /dev/null -s -w "%{http_code}" --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST:$ELASTICSEARCH_PORT/_opendistro/_ism/policies/${PREFIX}datahub_usage_event_policy") -eq 404 ]
-  then
-    echo -e "\ncreating datahub_usage_event_policy"
-    sed -e "s/PREFIX/${PREFIX}/g" /index/usage-event/aws_es_ism_policy.json | tee -a /tmp/aws_es_ism_policy.json
-    curl -XPUT --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST:$ELASTICSEARCH_PORT/_opendistro/_ism/policies/${PREFIX}datahub_usage_event_policy" -H 'Content-Type: application/json' --data @/tmp/aws_es_ism_policy.json
-  else
-    echo -e "\ndatahub_usage_event_policy exists"
-  fi
-  if [ $(curl -o /dev/null -s -w "%{http_code}" --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST:$ELASTICSEARCH_PORT/_template/${PREFIX}datahub_usage_event_index_template") -eq 404 ]
-  then
-    echo -e "\ncreating datahub_usage_event_index_template"
-    sed -e "s/PREFIX/${PREFIX}/g" /index/usage-event/aws_es_index_template.json | tee -a /tmp/aws_es_index_template.json
-    curl -XPUT --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST:$ELASTICSEARCH_PORT/_template/${PREFIX}datahub_usage_event_index_template" -H 'Content-Type: application/json' --data @/tmp/aws_es_index_template.json
-    curl -XPUT --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST:$ELASTICSEARCH_PORT/${PREFIX}datahub_usage_event-000001"  -H 'Content-Type: application/json' --data "{\"aliases\":{\"${PREFIX}datahub_usage_event\":{\"is_write_index\":true}}}"
-  else
-    echo -e "\ndatahub_usage_event_index_template exists"
+  # this fixes the case when datahub_usage_event was created by GMS before datahub_usage_event-000001
+  USAGE_EVENT_STATUS=$(curl -o /dev/null -s -w "%{http_code}\n" --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_URL/${PREFIX}datahub_usage_event")
+  if [ $USAGE_EVENT_STATUS -eq 200 ]; then
+    USAGE_EVENT_DEFINITION=$(curl -s --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_URL/${PREFIX}datahub_usage_event")
+    # the definition is expected to contain "datahub_usage_event-000001" string
+    if [[ $USAGE_EVENT_DEFINITION != *"datahub_usage_event-000001"* ]]; then
+      # ... if it doesn't, we need to drop it
+      echo -e "\n>>> deleting invalid datahub_usage_event ..."
+      curl -s -XDELETE --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_URL/${PREFIX}datahub_usage_event"
```

Comment:
Deleting an index in the setup job seems dangerous. We need to be EXTREMELY careful that this doesn't affect existing deployments. Can you explain a bit more why this drop is necessary?

Reply:
Will analytics necessarily _not_ work if the conditions is met to enter this block?

Reply:
Yes, if the index was previously created with `USE_AWS_ELASTICSEARCH` not set, and then later switched to `true`, Analytics tab doesn't work and no analytics events get ever recorded. (The name matches, but the contents do not.) The only solution (and the core fix of this PR) is to drop the wrong index and recreate it anew using `aws_es_usage_event.json` as source.

This block is executed only if `USE_AWS_ELASTICSEARCH=true` (we are on AWS) and only if there is an existing `datahub_usage_event ` index which doesn't contain the AWS-specific part — indicating the case described above.

What could go wrong? Maybe if the name of the index is later changed and this condition is not adjusted appropriately. Let me extract it to a constant to be safe and add some more comments in the code.

Reply:
Dropping here is probably too harsh here, we have an alias. If it's a mapping that needs to be fixed it should use the reindex API rather than a full drop of the index.